File: android/src/main/kotlin/dev/steenbakker/flutter_ble_central/handlers/StateChangedHandler.kt
package dev.steenbakker.flutter_ble_central.handlers

import android.os.Handler
import android.os.Looper
import dev.steenbakker.flutter_ble_central.models.CentralState
import io.flutter.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

class StateChangedHandler(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) : EventChannel.StreamHandler {
    private val tag: String = "BLE Central state "

    private var eventSink: EventChannel.EventSink? = null

    var state = CentralState.unknown

    private val eventChannel = EventChannel(
        flutterPluginBinding.binaryMessenger,
        "dev.steenbakker.flutter_ble_central/ble_state_changed"
    )

    init {
        eventChannel.setStreamHandler(this)
    }

    fun publishState(state: CentralState) {
        Log.i(tag, state.name)
        this.state = state
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(state.ordinal)
        }
    }

    override fun onListen(event: Any?, eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink
        publishState(state)
    }

    override fun onCancel(event: Any?) {
        this.eventSink = null
    }
}

File: darwin/flutter_ble_central/Sources/flutter_ble_central/handlers/StateChangedHandler.swift
//
//  StateChangedHandler.swift
//  flutter_ble_central
//
//  Created by Julian Steenbakker on 28/03/2022.
//

import Foundation
import CoreBluetooth

#if os(iOS)
  import Flutter
  import UIKit
#else
  import FlutterMacOS
#endif


public class StateChangedHandler: NSObject, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?

    var legacyState: CBManagerState = CBManagerState.unknown

    private let eventChannel: FlutterEventChannel

    init(registrar: FlutterPluginRegistrar) {
#if os(iOS)
        let messenger = registrar.messenger()
#else
        let messenger = registrar.messenger
#endif
        eventChannel = FlutterEventChannel(name: "dev.steenbakker.flutter_ble_peripheral/ble_state_changed",
                                           binaryMessenger: messenger)
        super.init()
        eventChannel.setStreamHandler(self)
    }

    func publishPeripheralState(state: CBManagerState) {
        if let eventSink = self.eventSink {
            eventSink(state.rawValue)
        }
    }
    
    func publishLegacyPeripheralState(state: CBManagerState) {
        self.legacyState = state
        if let eventSink = self.eventSink {
            eventSink(state.rawValue)
        }
    }

    public func onListen(withArguments arguments: Any?,
                         eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}
